File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        int top = 0, bottom = matrix.length-1, left = 0, right = matrix[0].length-1;
        while(left<=right && top <= bottom){
            for(int i=left;i<=right;i++){
                list.add(matrix[top][i]);
            }
            top++;
            for(int i=top;i<=bottom;i++){
                list.add(matrix[i][right]);
            }
            right--;
            if(top<=bottom){
                for(int i=right;i>=left;i--){
                    list.add(matrix[bottom][i]);
                }
                bottom--;
            }
            
            if(left<=right){
                for(int i=bottom;i>=top;i--){
                    list.add(matrix[i][left]);
                }
                left++;
            }
        }
        return list;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let m = matrix.length;
    let n = matrix[0].length;
    let array = [];
    let top = 0, left = 0, bottom = m-1, right = n-1;
    while(left <= right && top <= bottom){
        for(let i = left;i <= right ;i++){
            array.push(matrix[top][i]);
        }
        top++;
        for(let i=top;i<=bottom;i++){
            array.push(matrix[i][right]);
        }
        right--;
        if(top<=bottom){
            for(let i=right;i>= left;i--){
               array.push(matrix[bottom][i]); 
            }
            bottom--;
        }
        if (left <= right) {
            for(let i=bottom;i>=top;i--){
                array.push(matrix[i][left]);
            }
            left++;
        }
    }
    
    return array;
};